service/src/main/java/org/apache/griffin/core/job/LivyTaskSubmitHelper.java

```diff
@@ -0,0 +1,238 @@
+    private AtomicInteger curConcurrentTaskNum = new AtomicInteger(0);
+    private String workerNamePre;
+    private RestTemplate restTemplate = new RestTemplate();
+    // queue for pub or sub
+    private BlockingQueue<Map<String, Object>> queue;
+    public static final int DEFAULT_QUEUE_SIZE = 20000;
+    private String uri;
+
+    @Autowired
+    private Environment env;
+
+    @PostConstruct
+    public void init() {
+        startWorker();
+        uri = env.getProperty("livy.uri");
+        logger.info("Livy uri : {}",uri);
+    }
+
+    public LivyTaskSubmitHelper() {
+        this.workerNamePre = "livy-task-submit-worker";
+    }
+
+    public void startWorker() {
+        queue = new LinkedBlockingQueue<>(DEFAULT_QUEUE_SIZE);
+        BatchTaskWorker worker = new BatchTaskWorker();
+        worker.setDaemon(true);
+        worker.setName(workerNamePre + "-" + worker.getName());
+        worker.start();
+    }
+
+    public void addTaskToWaitingQueue(Map<String, Object> t) {
+        if (t == null) {
+            logger.warn("task is blank, workerNamePre: {}", workerNamePre);
+            return;
+        }
+
+        if (queue.remainingCapacity() <= 0) {
+            logger.warn("task is discard, workerNamePre: {}, task: {}", workerNamePre, JSON.toJSON(t));
+            return;
+        }
+
+        queue.add(t);
+
+        logger.info("add_task_to_waiting_queue_success, workerNamePre: {}, task: {}", workerNamePre, JSON.toJSON(t));
+    }
+
+    /**
+     * Consumer thread
+     */
+    class BatchTaskWorker extends Thread {
+        public void run() {
+            long insertTime = System.currentTimeMillis();
+
+            // Keep sequential execution within a limited number of tasks
+            while (true) {
+                try {
+                    if (curConcurrentTaskNum.get() < getMaxConcurrentTaskCount()
+                            && (System.currentTimeMillis() - insertTime) >= getBatchIntervalSecond() * 1000) {
+                        Map<String, Object> task = queue.take();
+                        doTask(task);
+                        insertTime = System.currentTimeMillis();
+                    }
+                } catch (Throwable e) {
+                    logger.error("Async_worker_doTask_failed, {}", workerNamePre + e.getMessage(), e);
+                }
+            }
+        }
+    }
+
+    public void increaseCurTaskNum(Long scheduleId) {
+        curConcurrentTaskNum.incrementAndGet();
+        if (scheduleId != null) taskAppIdMap.put(scheduleId, 1);
+    }
+
+    //Remove tasks after job status updates
+    public void decreaseCurTaskNum(Long scheduleId) {
+        if (scheduleId != null && taskAppIdMap.containsKey(scheduleId)) {
+            curConcurrentTaskNum.decrementAndGet();
+            taskAppIdMap.remove(scheduleId);
+        }
+    }
+
+    /**
+     * Submit a task to Livy and concurrent TaskNum++
+     */
+    protected void doTask(Map<String, Object> livyConfMap) {
```

Comment:
Looks like this is duplicating "post2Livy", probably it's better to merge it together.